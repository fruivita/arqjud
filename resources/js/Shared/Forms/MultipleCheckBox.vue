<!--
    MÃºltiplos input do tipo checkbox.

    @see https://vuejs.org/guide/introduction.html
    @see https://tailwindcss.com/docs
    @see https://inertiajs.com/
    @see https://kevinbigelow.medium.com/create-multiple-custom-checkboxes-bound-to-the-same-array-using-vue-3-and-the-composition-api-cbcdfe6216b7
 -->

<script setup>
import CheckBox from '@/Shared/Forms/CheckBox.vue';

const props = defineProps({
    disabled: { type: Boolean, default: false },
    value: { type: Array, required: true },
    opcoes: {
        type: Array,
        required: true,
        validator: (value) => {
            const hasNomeKey = value.every((option) => Object.keys(option).includes('nome'));
            const hasIdKey = value.every((option) => Object.keys(option).includes('id'));
            return hasNomeKey && hasIdKey;
        },
    },
});

const emit = defineEmits(['update:value']);

const check = (optionId, checked) => {
    if (props.disabled) {
        return;
    }

    // copy the value Array to avoid mutating props
    let updatedValue = [...props.value];
    // remove name if checked, else add name
    if (checked) {
        updatedValue.push(optionId);
    } else {
        updatedValue.splice(updatedValue.indexOf(optionId), 1);
    }

    emit('update:value', updatedValue);
};
</script>

<template>
    <div class="flex flex-col items-start space-y-3">
        <CheckBox
            v-for="opcao in opcoes"
            :id="opcao.nome"
            :key="opcao"
            :checked="value.includes(opcao.id)"
            :disabled="disabled"
            :label="opcao.nome"
            @update:checked="check(opcao.id, $event)"
        />
    </div>
</template>
