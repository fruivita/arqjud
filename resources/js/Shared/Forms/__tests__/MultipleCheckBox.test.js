/**
 * Testes para o componente MultipleCheckBox.
 *
 * @link https://vuejs.org/guide/scaling-up/testing.html
 * @link https://test-utils.vuejs.org/guide/
 * @link https://vitest.dev/
 */

import MultipleCheckBox from '@/Shared/Forms/MultipleCheckBox.vue';
import { mount } from '@vue/test-utils';
import { beforeEach, describe, expect, test } from 'vitest';

let mountFunction;

beforeEach(() => {
    mountFunction = (options = {}) => {
        return mount(MultipleCheckBox, { ...options });
    };
});

describe('MultipleCheckBox', () => {
    test('propriedades estão definidas', () => {
        expect(MultipleCheckBox.props).toMatchObject({
            disabled: { type: Boolean, default: false },
            value: { type: Array, required: true },
            opcoes: { type: Array, required: true },
        });
    });

    test('opções devem possuir chave nome e id', () => {
        const validator = MultipleCheckBox.props.opcoes.validator;

        expect(validator([{ foo: 1, loren: 'foo' }])).toBe(false);
        expect(validator([{ id: 1, nome: 'foo' }])).toBe(true);
    });

    test('renderiza o componente respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    value: [],
                    opcoes: [
                        { id: 1, nome: 'foo' },
                        { id: 2, nome: 'bar' },
                        { id: 3, nome: 'baz' },
                    ],
                },
            }).html()
        ).toMatchSnapshot();
    });

    test('renderiza o componente desabilitado respeitando o snapshot', () => {
        expect(
            mountFunction({
                props: {
                    disabled: true,
                    value: [],
                    opcoes: [
                        { id: 1, nome: 'foo' },
                        { id: 2, nome: 'bar' },
                        { id: 3, nome: 'baz' },
                    ],
                },
            }).html()
        ).toMatchSnapshot();
    });
});
