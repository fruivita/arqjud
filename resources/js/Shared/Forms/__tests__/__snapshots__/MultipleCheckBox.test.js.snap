// Vitest Snapshot v1

exports[`MultipleCheckBox > renderiza o componente desabilitado respeitando o snapshot 1`] = `"<div class=\\"flex flex-col items-start space-y-3\\"><label for=\\"foo\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"foo\\" disabled=\\"\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>foo</span></label><label for=\\"bar\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"bar\\" disabled=\\"\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>bar</span></label><label for=\\"baz\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"baz\\" disabled=\\"\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>baz</span></label></div>"`;

exports[`MultipleCheckBox > renderiza o componente respeitando o snapshot 1`] = `"<div class=\\"flex flex-col items-start space-y-3\\"><label for=\\"foo\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"foo\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>foo</span></label><label for=\\"bar\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"bar\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>bar</span></label><label for=\\"baz\\" class=\\"inline-flex select-none items-center space-x-2\\" dusk=\\"label\\"><input id=\\"baz\\" class=\\"h-5 w-5 accent-primaria-500\\" type=\\"checkbox\\"><span>baz</span></label></div>"`;
